eclipse/che
# use precreated labeled namespace for workspaces

Signed-off-by: Michal Vala <mvala@redhat.com>



### What does this PR do?
Allow using precreated namespaces labeled with proper labels. The labels are configurable with `che.infra.kubernetes.namespace.labels` property.

### Documentation PR
https://github.com/eclipse/che-docs/pull/1643

TODO:
 - [x] Make labeled namespace as a valid namespace for creating new workspace (TODO in KubernetesNamespaceFactory.java)
 - [x] Javadocs

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/17842


### How to test this PR?

1. create namespace with labels `app.kubernetes.io/part-of=che.eclipse.org,app.kubernetes.io/component=workspace,che.eclipse.org/username=<username>` and make sure your user has access to it (`admin` role)
  1. On OpenShift with OAuth, it should be enough just to create RoleBinding for the user and `admin` role in pre-created project.
  1. On Kubernetes, you will also need to add list namespaces clusterrole permission to `che` serviceaccount.
1. login to Che with the user
1. go to create new namespace page
1. you should see the namespace name in the dropdown "Namespace"
1. when you use factory to create a namespace, this labeled namespace will be used
1. when you label more namespaces, all will be shown in dashboard. If you use factory with more labeled namespaces, the first one will be used (should be alphabetical order, but not guaranteed)


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
